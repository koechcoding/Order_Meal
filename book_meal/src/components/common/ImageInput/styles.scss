.preview,
canvas {
    cursor: pointer;
    width: 250px;
    margin-left: calc(50% - 125px);
    text-align: center;
}
.prompt {
    margin: auto;
    width: 300px;
    max-width: 480px;
    background-color: #eee;
    color: #555;
    font-size: 28px;
    text-align: center;
    padding: 10px;
    cursor: pointer;
}
img.prompt {
    background-color: white;
    padding-bottom: 0;
}
.hint {
    text-align: center;
    margin-top: 0;
}
.inner {
    margin: 10px;
    padding: 10px;
    padding-top: 40px;
    padding-bottom: 40px;
    border: 5px dashed #ccc;
}