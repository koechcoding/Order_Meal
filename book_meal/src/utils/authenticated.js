export default function authenticated(){
    try{
        return JSON.parse(localStorage.getItem('user'));
    } catch(err){
        
    }
}