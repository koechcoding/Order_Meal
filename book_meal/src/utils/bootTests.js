import Enzyme from "enzyme";
import Adapter from "enzyme-adapter-react-16";
Enzyme.configure({ adapter: new Adapter()});

global.window = {}
import localStorage from 'mock-local-Storage';
window.localStorage = global.localStorage

export default Enzyme;
