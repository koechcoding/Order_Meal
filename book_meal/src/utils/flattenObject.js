export default function flattenObject(obj){
    let result = {}
    for (const key in obj){
        if((typeof obj[key])=== 'object' &&  obj[key]  !== null) {
            let flattened = flattenObject(obj[key])
            for(const key2 in flattened){
                result[key + "." + key2] = flattened[key2];
            }
        }else{
            result[key] = obj[key];
        }
    }
    return result;
}