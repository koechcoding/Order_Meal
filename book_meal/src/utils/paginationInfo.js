export default function paginationInfo(page){

    if(!page){
        return{}
    }

    let currentPage = 1
    if(page.next_page)
         currentPage = page.next_page -1;
    else if (page.prev_page)
        currentPage = page.prev_page + 1;

    let pageCount = Math.ceil(page.total / page.per_paage);
    if(Number.isNaN(pageCount) || pageCount === 0)
      pageCount = 1; 

    return {
        pageCount,
        currentPage,
        total: page.total,
        perPage: page.per_page,
        hasNext: page.has_next,
        hasPrev: page.has_prev,
        nextPage: page.next_page,
        prevPage: page.prev_page,
        currentCount: page.current_count,
    };
}