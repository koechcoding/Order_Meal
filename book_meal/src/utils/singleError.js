const singleError = (errorPayload)=>{
    if(errorPayload && errorPayload.data){
        const response = errorPayload.data;
        //alternative to Object.keys...
        for(let errorKey in response.errors){
            return response.error[errorKey]
        }
        if(response.message){
            return response.message;
        }
    }
    return 'Unknown error. Please try again!';
}
export default singleError;