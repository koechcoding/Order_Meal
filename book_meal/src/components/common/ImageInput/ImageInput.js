import React from 'react';
import PropTypes from 'prop-types';
import './styles.css';

class ImageInput extends React.Component {
    constructor(props){
        super(props);
        let prefillSelected = false;
        if(props.prefill && props.prefill !== '#'){
            prefillSelected = true;
        }
        this.state = {
            size: 1,
            fileName: '',
            fileSize: '',
            fileModified: '',
            imageSelected: false,
            prefillSelected,
        }
        this.canvasRef = React.createRef();
        this.containerRef = React.createRef();
        this.fileInputRef = React.createRef();
    }
    onRemove =(e)=>{
        this.setState({
            ...this.state,
            fileName: '';
            fileSize: '',
            fileModified: '',
            imageSelected: false,
        });
        this.fileInputRef.current.value = '';
        this.props.onImageRemoved();
    }
    onRemovePrefill= (e)=>{
        this.setState({
            ...this.state,
            fileName: '',
            fileSize: '',
            fileModified: '',
            imageSelected: false,
            prefillSelected: false,
        });
        this.props.onPrefillRemoved();
    }
    onSelectImage = (e)=>{
        this.fileInputRef.current.click();
    }

    onFileChange =(e)=>{
        const files = e.target.files
        if(!files.length){
            return
        }
        const {
            size,
            fileName,
            fileSize,
            fileModified
        } = this.state;
        if(files[0].size <= 0 || files[0].size > size *1024 * 1024){
            this.setState({
                ...this.state,
                error: 'Image file is too large. Please select a smaller image'
            });
            return
        }
       
        //if not unique file return
        if(files[0].name === fileName && files[0].size === fileSize && fileModified === files[0].lastModified){
            return
        }
        //ensure received file is image
        if(files[0].type.substr(0, 6) !== 'image/'){
            return
        }

        this.setState({
            ...this.state,
            file: files[0],
            fileName: files[0].name,
            fileSize: files[0].size,
            fileModified: files[0].lastModified,
            fileType: files[0].type,
            imageSelected: true,
        });
        this.loadImage(files[0]);
    }

    loadImage = (file)=>{
        const reader = new FileReader()
        reader.onload =(e)=>{
            let image = new Image();
            image.onload =()=>{
                this.drawImage(image);
            }
            image.src = e.target.result;
            this.props.onImageAdded(e.target.result);
        }
        reader.readAsDataURL(file)
    }

    drawImage=(image)=>{

        const width = 300;
        const height = 300;
        let previewRatio = width/height

        const imageRatio = image.width / image.height
        const containerWidth = this.containerRef.current.clientWidth
        const previewWidth = Math.min(containerWidth, width)
        const previewHeight = previewWidth / previewRatio

        const pixelRatio = Math.round(window.devicePixelRatio || window.screen.deviceXDPI / window.screen.logicalXDPI )

        let offsetX = 0
        let offsetY = 0
        let scaleWidth = previewWidth
        let scaleHeight = previewHeight / previewRatio

        //crop the image
        if(imageRatio >= previewRatio){
            scaledWidth = scaledHeight * imageRatio;
            offsetX = (previewWidth - scaleWidth) / 2;
        }else {
            scaleHeight = scaleWidth / imageRatio;
            offsetY = (previewHeight - scaleWidth) / 2;
        }

        const canvas = this.canvasRef.current
        const context = canvas.getContext('2d')
        canvas.style.background = 'none'
        canvas.width = previewWidth * pixelRatio
        canvas.height = previewHeight * pixelRatio
        canvas.setTransform(1, 0, 0, 1, 0, 0)
        context.drawImage(image,
            offsetX * pixelRatio,
            offsetY * pixelRatio,
            scaleWidth * pixelRatio,
            scaledHeight * pixelRatio
        )
    }
    render(){

    }
}