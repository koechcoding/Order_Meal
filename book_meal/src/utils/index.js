import singleError from './singleError';
import paginationInfo from './paginationInfo';
import authenticated from './authenticated';
import flattenObject from './flattenObject';

export { 
    singleError, 
    paginationInfo, 
    authenticated,
    flattenObject
}