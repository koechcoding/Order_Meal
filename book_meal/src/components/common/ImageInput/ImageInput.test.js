import React from 'react';
import { shallow, mount }  from 'enzyme';
import 'src/utils/bootTests';
import ImageInput from './ImageInput';
import Adapter from 'enzyme-adapter-react-16';

describe('<ImageInput />', () => {
    it('renders correctly', () => {
        const wrapper = shallow(<ImageInput/>);
        expect(wrapper).toMatchSnapshot();
        expect(wrapper.find('.inner').text()).toEqual('Click to select image from your files.');
    });
    it('renders prefill image', () => {
        const wrapper = shallow(<ImageInput prefill={'https://plachold.it/10x10'}/>);
        expect(wrapper.find('.preview').length).toBe(1);
    });
    it('removes prefill image', () => {
        const wrapper = shallow(<ImageInput onPrefillRemoved={() => {}} prefill={'https://plachold.it/10x10'}/>);
        expect(wrapper.find('.preview').length).toBe(1);
        wrapper.find('.btn-danger').simulate('click');
        expect(wrapper.find('.preview').length).toBe(0);
    });
    it('selects image', () => {
        const wrapper = mount(<ImageInput onPrefillRemoved={() => {}} prefill={'https://plachold.it/10x10'}/>);
        wrapper.find('input').simulate('change', { 
            target: {
                files: [new File(['abc'], 'file.png', {type: 'image/png'})]
            }
        });
        expect(wrapper.state('fileName')).toEqual('file.png');
    });
    it('removes selected image', () => {
        const wrapper = mount(<ImageInput onPrefillRemoved={() => {}} onImageRemoved={() => {}}/>);
        wrapper.find('input').simulate('change', { 
            target: {
                files: [new File(['abc'], 'file.png', {type: 'image/png'})]
            }
        });
        expect(wrapper.state('fileName')).toEqual('file.png');
        wrapper.find('.remove').simulate('click');
        expect(wrapper.state('fileName')).toEqual('');
    });
});