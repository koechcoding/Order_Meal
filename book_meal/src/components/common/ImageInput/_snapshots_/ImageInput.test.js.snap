// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ImageInput /> renders correctly 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <ImageInput />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <input
          className="d-none"
          onChange={[Function]}
          type="file"
/>,
        <div
          className="prompt"
          onClick={[Function]}
>
          <div
                    className="inner"
          >
                    Click to select image from your files.
          </div>
</div>,
      ],
    },
    "ref": Object {
      "current": null,
    },
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "className": "d-none",
          "onChange": [Function],
          "type": "file",
        },
        "ref": Object {
          "current": null,
        },
        "rendered": null,
        "type": "input",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": <div
            className="inner"
>
            Click to select image from your files.
</div>,
          "className": "prompt",
          "onClick": [Function],
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Click to select image from your files.",
            "className": "inner",
          },
          "ref": null,
          "rendered": "Click to select image from your files.",
          "type": "div",
        },
        "type": "div",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <input
            className="d-none"
            onChange={[Function]}
            type="file"
/>,
          <div
            className="prompt"
            onClick={[Function]}
>
            <div
                        className="inner"
            >
                        Click to select image from your files.
            </div>
</div>,
        ],
      },
      "ref": Object {
        "current": null,
      },
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "className": "d-none",
            "onChange": [Function],
            "type": "file",
          },
          "ref": Object {
            "current": null,
          },
          "rendered": null,
          "type": "input",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": <div
              className="inner"
>
              Click to select image from your files.
</div>,
            "className": "prompt",
            "onClick": [Function],
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "Click to select image from your files.",
              "className": "inner",
            },
            "ref": null,
            "rendered": "Click to select image from your files.",
            "type": "div",
          },
          "type": "div",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;